### ISSUE TITLE
fix(widgetbook): using `flutter_screenutil`

### ISSUE THREAD

**Issue by kalthoum23** [widgetbook/widgetbook#412]
**Description**
i'm working on a widgetbook project , i'm displaying my project widgets using themes and screenUtils
When adding the screenUtils to display widgets with theme , widgets are not getting the right size , scaling is not working as showen in the captures 

**code exemple :**

```dart 
Widget buildStorybook(BuildContext context) {

    return ScreenUtilInit(
       designSize: const Size(360, 760),
        useInheritedMediaQuery: true,
        minTextAdapt: true,
        builder: (context, child) {
          return Widgetbook.material(
            deviceFrameBuilder: (context, device, frame, orientation, child) {
              if (_currentDevice != device) {
                _currentDevice = device;
              }
              return defaultDeviceFrameBuilder(
                context,
                device,
                frame,
                orientation,
                child,
              );
            },
            devices: _devices,
            textScaleFactors: [1],
            themes: [
//themes of the app using screenUtils dimensions (.h,.w...)
              WidgetbookTheme(
                name: 'Light',
                data: LightTheme.themeData
              ),
              WidgetbookTheme(
                  name: 'dark',
                  data:DarkTheme.themeData
              ),
            ],
            categories: [
              buttonWidgetBook,
             
            ],
            appInfo: AppInfo(name: 'WidgetBook App'),
            appBuilder: (context, child) {
              return child;
            });
        });
  }

```

**Expected Behavior**
<img width="657" alt="image" src="https://user-images.githubusercontent.com/73825503/209564193-28264289-9e99-4338-b30b-218e13d2fc7b.png">

**Result**

<img width="657" alt="image" src="https://user-images.githubusercontent.com/73825503/209564361-65efebd4-352f-4508-97dc-c55b89b354bb.png">

**Comment by Xazin** [comment]
Hey @kalthoum23 , I'm curious about this, I have not used `ScreenUtil` myself though.

Is it possible for you to share with me the code for the `buttonWidgetBook` (not only the variable but the Widget itself).

If you do please tag/reply to me, and I will take a look to see if there is a way to take a different approach that could make it work for you.

**Comment by shigomany** [comment]
You will try use this [fork](https://github.com/widgetbook/widgetbook/pull/417)

**Comment by straiser** [comment]
I've got the same issue. It seems that widgetbook gets screen size from whole widgetbook, not only from device preview. Because I also use screen utils in theme,  I can't use mediaquery etc. to fool screen_utils.  Is there any workaround for this?

**Comment by kalthoum23** [comment]
Hi @Xazin , when i declare my custom theme 
```dart 
  themes: [
          WidgetbookTheme(name: 'Light', data: LightTheme.themeData),
          WidgetbookTheme(name: 'dark', data: DarkTheme.themeData),
        ],
```
where i use screenUtils i get wrong sizes of widgets.

As a temporary solution @straiser, I forced the theme in each component 

```dart 
     WidgetbookUseCase(
        name: "Secondary Text Button",
        builder: (context) => Theme(
              data: context.knobs.options(label: "Theme", options: [
                Option(label: "Light", value: LightTheme.themeData),
                Option(label: "Dark", value: DarkTheme.themeData)
              ]),
              child: Scaffold(
                body: Center(
                  child: BasicButton(
                    onPressed: context.knobs.options(label: "Status", options: [
                      Option(label: "enabled", value: () {}),
                      const Option(label: "disabled", value: null),
                    ]),
                    text: context.knobs
                        .text(initialValue: "Custom button", label: "title"),
                    
                  ),
                ),
              ),
            )),
```

**Comment by Groseuros** [comment]
Facing same issue with `ScreenUtil` package. I've set designSize to `const Size(1920, 1080)` for now to fix to work with `Widgetbook` but I'm really looking for a fix to this

**Comment by Rohithgilla12** [comment]
@Groseuros Update to `widgetbook 3`, it works.

**Comment by Groseuros** [comment]
@Rohithgilla12 I'm already using WidgetBook 3. It works with ScreenUtil but sizes / ratio are wrong because it uses the WidgetBook app constraints and not the mobile layout constraints.

**Comment by Rohithgilla12** [comment]
@Groseuros Oh yes, sorry.
I did check, a few widgets are breaking.


@jenshor please help 😄

**Comment by jenshor** [comment]
@Rohithgilla12 can you share your 'appBuilder' (if customized) and use-case builder code? 

I might be able to look into this later.

**Comment by Rohithgilla12** [comment]
@jenshor Nothing special in the builder, also can't use `useCaseBuilder` in widgetbook 3, I didn't find the builder exported anywhere.

Using a one almost similar to this.
https://github.com/widgetbook/widgetbook/blob/0f65e36a0a63a2585564696617f2d49a122b0add/examples/custom_theme_example/widgetbook/main.widgetbook.dart#L82

**Comment by jenshor** [comment]
@Rohithgilla12 Ok, I'll have a look. the `useCaseBuilder` (and the other builders) has been refactored and merged with the `appBuilder` as we think the API was too complicated. 

You can find a preview documentation regarding the new `appBuilder` here: https://docs.widgetbook.io/~518/getting-started/configuration#customize-how-use-cases-are-previewed

I have not used `ScreenUtil` yet, but I'll look into it now. If it's `MediaQuery` based it should work properly but setting it up with the `appBuilder` might require the correct `BuildContext` to be set. 

Examples of the default `materialAppBuilder` (and others) can be found here: https://github.com/widgetbook/widgetbook/blob/main/packages/widgetbook/lib/src/builder/functions/app_builder.dart

**Comment by Rohithgilla12** [comment]
@jenshor I think I got it to work, thanks!

I can create an example repo to demonstrate it or if it's okay can make PR to the repo too by adding an example.

**Comment by jenshor** [comment]
@Rohithgilla12, I just created the PR. Feel free to have a look and let me know if and how I can improve the example. The most relevant file is the [main.dart file](https://github.com/widgetbook/widgetbook/pull/540/files#diff-4b2e310b4c1cccb795160b2300e6cba0f048d92fe1206654c848aef8219f6aad)

**Comment by Rohithgilla12** [comment]
@jenshor Btw thanks for the example.

The only issue I found is sometimes, the images are breaking.

If you want me to share my screen/code and explain let me know. We could connect and discuss about it 😄

**Comment by jenshor** [comment]
Yes, please share your code/screen or post a reproducible example regarding the images so I can follow what's going wrong. I'm also happy to connect and solve this problem together 👍🏻

**Comment by Rohithgilla12** [comment]
@jenshor Sure, can we do a small call?
https://zcal.co/rohithgilla/30min


If the slots doesn't look good, let me know I can open more.

Thanks

**Comment by Rohithgilla12** [comment]
@jenshor Hello as discussed here is the reproducible example

https://github.com/Rohithgilla12/widgetbook_screenutil_bug

I have added a readme on how to reproduce it and what might be the issue.

![https://getkap-gifs.s3.ap-south-1.amazonaws.com/Kapture+2023-03-20+at+09.19.24.gif](https://getkap-gifs.s3.ap-south-1.amazonaws.com/Kapture+2023-03-20+at+09.19.24.gif)

TL;DR
It happens on a widget re-render.

**Comment by jenshor** [comment]
@Rohithgilla12 Ok. Thanks for the reproducible code. I'm looking into it now.

**Comment by jenshor** [comment]
@Rohithgilla12 what's the expected output? The image before or after clicking the button? 

I already found the problem, but not yet sure how to solve it.

**Comment by Rohithgilla12** [comment]
> @Rohithgilla12 what's the expected output? The image before or after clicking the button? 
> 
> 
> 
> I already found the problem, but not yet sure how to solve it. 

@jenshor 
The size of the image should remain the same, I think it only happens with width and not heights.

**Comment by jenshor** [comment]
It probably happens with both. I've found a solution to the problem already. I'll see if there is a better and cleaner approach. 

To me, it looks like a `ScreenUtil` problem as it uses an inherited `context` to get the `Navigator` via `Navigator.of(context)`. From this navigator, the screen_util package gets the `context`. The code basically does `Navigator.of(context).context`. The `context` from the `Navigator` might not be the inherited one. Sounds confusing so
[see the relevant code of the `screen_util` package](https://github.com/OpenFlutter/flutter_screenutil/blob/master/lib/src/screen_util.dart#L99)

**Comment by jenshor** [comment]
@Rohithgilla12 can you try this custom `appBuilder` function?
I'll check with the authors of the `flutter_screenutil` package regarding their implementation.

```
Widget appBuilderForScreenUtil(
  BuildContext context,
  Widget child,
) {
  final frameBuilder = context.frameBuilder;
  final theme = context.theme<ThemeData>();

  final builder = Builder(
    builder: (context) {
      return MaterialApp(
        useInheritedMediaQuery: true,
        theme: theme,
        locale: context.localization?.activeLocale,
        supportedLocales: context.localization?.locales ??
            const <Locale>[
              Locale('en', 'US'),
            ],
        localizationsDelegates: context.localization?.localizationsDelegates,
        debugShowCheckedModeBanner: false,
        home: ScreenUtilInit(
          minTextAdapt: true,
          designSize: const Size(375, 812),
          // Enable this property so the [MediaQuery] of the [FrameAddon] is used.
          useInheritedMediaQuery: true,
          builder: (context, child) {
            return Scaffold(
              body: MediaQuery(
                data: MediaQuery.of(context).copyWith(
                  textScaleFactor: context.textScale,
                ),
                child: child!,
              ),
            );
          },
          child: child,
        ),
      );
    },
  );

  // Note, that the ScreenUtilInit [Widget] only works properly iff the
  // [FrameAddon] is active!
  return frameBuilder == null
      ? builder
      : frameBuilder(
          context,
          builder,
        );
}```

**Comment by Rohithgilla12** [comment]
@jenshor Thanks, it works!
I have also updated the repo with the workaround commit!

https://github.com/Rohithgilla12/widgetbook_screenutil_bug/commit/72135ddfb32533a6255fd05a80a46265189e8705

**Comment by jenshor** [comment]
Added an example with the linked PR. Still waiting for https://github.com/OpenFlutter/flutter_screenutil/issues/466 to be worked on
